File: 983/983.kt
// Adapted from https://raw.githubusercontent.com/cirosantilli/project-euler-solutions/refs/heads/master/solvers/983.py

val OFF = 1 shl 15
val SHIFT = 17

fun encodeCoord(x: Int, y: Int): Int {
    return ((x + OFF) shl SHIFT) or (y + OFF)
}

fun circlePoints(m: Int): List<Pair<Int, Int>> {
    val lim = Math.sqrt(m.toDouble()).toInt()
    val pts = mutableListOf<Pair<Int, Int>>()
    for (x in -lim..lim) {
        val y2 = m - x * x
        if (y2 < 0) continue
        val y = Math.sqrt(y2.toDouble()).toInt()
        if (y * y == y2) {
            pts.add(Pair(x, y))
            if (y != 0) pts.add(Pair(x, -y))
        }
    }
    return pts
}

fun computeOppositePairs(points: List<Pair<Int, Int>>): List<Pair<Pair<Int, Int>, Pair<Int, Int>>> {
    val pairs = mutableListOf<Pair<Pair<Int, Int>, Pair<Int, Int>>>()
    val used = mutableSetOf<Int>()
    val sorted = points.sortedWith(compareBy({ it.first }, { it.second }))
    for (v in sorted) {
        val vk = encodeCoord(v.first, v.second)
        if (vk in used) continue
        val w = Pair(-v.first, -v.second)
        used.add(vk)
        used.add(encodeCoord(w.first, w.second))
        pairs.add(Pair(v, w))
    }
    return pairs
}

fun countAntipodalPairs(m: Int): Int {
    var x = m
    while (x % 2 == 0) x /= 2
    var product = 1
    var p = 3
    while (p * p <= x) {
        if (x % p == 0) {
            var exponent = 0
            while (x % p == 0) { x /= p; exponent++ }
            if (p % 4 == 1) {
                product *= exponent + 1
            } else if ((exponent and 1) != 0) {
                return 0
            }
        }
        p += 2
    }
    if (x > 1) {
        if (x % 4 == 1) product *= 2
        else if (x % 4 == 3) return 0
    }
    return 2 * product
}

fun buildDisplacementSet(points: List<Pair<Int, Int>>): Set<Int> {
    val deltas = mutableSetOf<Int>()
    for ((ax, ay) in points) {
        for ((bx, by) in points) {
            if (ax != bx || ay != by) {
                deltas.add(encodeCoord(ax - bx, ay - by))
            }
        }
    }
    return deltas
}

fun checkFourVectorPrune(selected: List<Pair<Int, Int>>, candidate: Pair<Int, Int>, deltas: Set<Int>): Boolean {
    if (selected.size < 3) return true
    val (vx, vy) = candidate
    val sel = selected.size
    for (i in 0 until sel - 2) {
        val (ax, ay) = selected[i]
        for (j in i + 1 until sel - 1) {
            val (bx, by) = selected[j]
            for (k in j + 1 until sel) {
                val (cx, cy) = selected[k]
                for (sa in intArrayOf(1, -1)) {
                    val x1 = vx + sa * ax
                    val y1 = vy + sa * ay
                    for (sb in intArrayOf(1, -1)) {
                        val x2 = x1 + sb * bx
                        val y2 = y1 + sb * by
                        var xv = x2 + cx
                        var yv = y2 + cy
                        if ((xv != 0 || yv != 0) && encodeCoord(xv, yv) in deltas) return false
                        xv = x2 - cx
                        yv = y2 - cy
                        if ((xv != 0 || yv != 0) && encodeCoord(xv, yv) in deltas) return false
                    }
                }
            }
        }
    }
    return true
}

fun buildEvenMasks(k: Int): List<Int> {
    val masks = mutableListOf<Int>()
    for (mask in 0 until (1 shl k)) {
        var bits = 0; var mm = mask
        while (mm != 0) { bits += mm and 1; mm = mm ushr 1 }
        if ((bits and 1) == 0) masks.add(mask)
    }
    return masks
}

fun computeCenters(vectors: List<Pair<Int, Int>>, masks: List<Int>): List<Pair<Int, Int>> {
    val centers = mutableListOf<Pair<Int, Int>>()
    for (mask in masks) {
        var x = 0; var y = 0; var mm = mask
        while (mm != 0) {
            val lsb = mm and -mm
            var idx = 0; var tmp = lsb; while (tmp > 1) { tmp = tmp ushr 1; idx++ }
            x += vectors[idx].first
            y += vectors[idx].second
            mm -= lsb
        }
        centers.add(Pair(x, y))
    }
    return centers
}

fun quickHarmonyCheck(centers: List<Pair<Int, Int>>, pts: List<Pair<Int, Int>>, n: Int): Boolean {
    val counts = mutableMapOf<Int, Int>()
    var harmonyCount = 0
    for ((cx, cy) in centers) {
        for ((vx, vy) in pts) {
            val key = encodeCoord(cx + vx, cy + vy)
            val cur = counts[key]
            if (cur == null) counts[key] = 1
            else if (cur == 1) { counts[key] = 2; harmonyCount++; if (harmonyCount > n) return false }
            else counts[key] = cur + 1
        }
    }
    return harmonyCount == n
}

fun strictCheck(centers: List<Pair<Int, Int>>, pts: List<Pair<Int, Int>>, n: Int): Boolean {
    val centerCodes = centers.map { (x, y) -> encodeCoord(x, y) }
    val centerSet = centerCodes.toHashSet()
    val tangentDiffs = pts.map { (x, y) -> encodeCoord(2 * x, 2 * y) }
    for (c in centerCodes) {
        for (d in tangentDiffs) {
            val other = c + d
            if (other in centerSet && c < other) return false
        }
    }
    val ptToCenters = mutableMapOf<Int, MutableList<Int>>()
    for ((idx, center) in centers.withIndex()) {
        val (cx, cy) = center
        for ((vx, vy) in pts) {
            val key = encodeCoord(cx + vx, cy + vy)
            ptToCenters.getOrPut(key) { mutableListOf() }.add(idx)
        }
    }
    val harmonyPts = ptToCenters.entries.filter { it.value.size >= 2 }.map { it.key }
    if (harmonyPts.size != n) return false
    val parent = IntArray(n) { it }
    val sz = IntArray(n) { 1 }
    fun find(xi: Int): Int {
        var xv = xi
        while (parent[xv] != xv) { parent[xv] = parent[parent[xv]]; xv = parent[xv] }
        return xv
    }
    fun union(a: Int, b: Int) {
        var ra = find(a); var rb = find(b)
        if (ra == rb) return
        if (sz[ra] < sz[rb]) { val tmp = ra; ra = rb; rb = tmp }
        parent[rb] = ra; sz[ra] += sz[rb]
    }
    for (key in harmonyPts) {
        val lst = ptToCenters[key]!!
        val base = lst[0]
        for (j in 1 until lst.size) union(base, lst[j])
    }
    val root = find(0)
    for (i in 1 until n) { if (find(i) != root) return false }
    return true
}

fun hasUnitCoord(points: List<Pair<Int, Int>>): Boolean {
    for ((x, y) in points) { if (Math.abs(x) == 1 || Math.abs(y) == 1) return true }
    return false
}

fun checkValidOrientedVectors(pairs: List<Pair<Pair<Int, Int>, Pair<Int, Int>>>, pts: List<Pair<Int, Int>>, k: Int, masks: List<Int>, n: Int): Boolean {
    val deltas = buildDisplacementSet(pts)
    val selected = mutableListOf<Pair<Int, Int>>()
    fun dfs(start: Int): Boolean {
        if (selected.size == k) {
            val centers = computeCenters(selected, masks)
            return quickHarmonyCheck(centers, pts, n) && strictCheck(centers, pts, n)
        }
        val needed = k - selected.size
        for (pi in start..pairs.size - needed) {
            val choices = if (selected.isEmpty()) listOf(pairs[pi].first) else listOf(pairs[pi].first, pairs[pi].second)
            for (vec in choices) {
                if (!checkFourVectorPrune(selected, vec, deltas)) continue
                selected.add(vec)
                if (dfs(pi + 1)) return true
                selected.removeAt(selected.size - 1)
            }
        }
        return false
    }
    return dfs(0)
}

fun searchMinRadiusSq(k: Int, mLimit: Int, filtered: Boolean): Int {
    val masks = buildEvenMasks(k)
    val n = 1 shl (k - 1)
    for (m in 1..mLimit) {
        val p = countAntipodalPairs(m)
        if (p < k) continue
        if (filtered && p != k && p != k + 2) continue
        val pts = circlePoints(m)
        if (filtered && !hasUnitCoord(pts)) continue
        val pairs = computeOppositePairs(pts)
        if (pairs.size != p) continue
        if (checkValidOrientedVectors(pairs, pts, k, masks, n)) return m
    }
    throw RuntimeException("No solution found up to m=$mLimit")
}

fun main() {
    val r1 = searchMinRadiusSq(2, 20, false)
    check(r1 == 1) { "Expected 1, got $r1" }
    val r2 = searchMinRadiusSq(3, 50, false)
    check(r2 == 5) { "Expected 5, got $r2" }
    println(searchMinRadiusSq(10, 20000, true))
}

File: 983/983.rs
// Adapted from https://raw.githubusercontent.com/cirosantilli/project-euler-solutions/refs/heads/master/solvers/983.py

use std::collections::{HashMap, HashSet};

const OFF: i32 = 1 << 15;
const SHIFT: i32 = 17;

fn encode_coord(x: i32, y: i32) -> i32 {
    ((x + OFF) << SHIFT) | (y + OFF)
}

fn circle_points(m: i32) -> Vec<(i32, i32)> {
    let lim = (m as f64).sqrt() as i32;
    let mut pts = Vec::new();
    for x in -lim..=lim {
        let y2 = m - x * x;
        if y2 < 0 { continue; }
        let y = (y2 as f64).sqrt() as i32;
        if y * y == y2 {
            pts.push((x, y));
            if y != 0 { pts.push((x, -y)); }
        }
    }
    pts
}

fn compute_opposite_pairs(points: &[(i32, i32)]) -> Vec<((i32, i32), (i32, i32))> {
    let mut pairs = Vec::new();
    let mut used = HashSet::new();
    let mut sorted = points.to_vec();
    sorted.sort();
    for v in &sorted {
        let vk = encode_coord(v.0, v.1);
        if used.contains(&vk) { continue; }
        let w = (-v.0, -v.1);
        used.insert(vk);
        used.insert(encode_coord(w.0, w.1));
        pairs.push((*v, w));
    }
    pairs
}

fn count_antipodal_pairs(m: i32) -> i32 {
    let mut x = m;
    while x % 2 == 0 { x /= 2; }
    let mut product = 1i32;
    let mut p = 3i32;
    while p * p <= x {
        if x % p == 0 {
            let mut exponent = 0;
            while x % p == 0 { x /= p; exponent += 1; }
            if p % 4 == 1 {
                product *= exponent + 1;
            } else if exponent & 1 != 0 {
                return 0;
            }
        }
        p += 2;
    }
    if x > 1 {
        if x % 4 == 1 { product *= 2; }
        else if x % 4 == 3 { return 0; }
    }
    2 * product
}

fn build_displacement_set(points: &[(i32, i32)]) -> HashSet<i32> {
    let mut deltas = HashSet::new();
    for &(ax, ay) in points {
        for &(bx, by) in points {
            if ax != bx || ay != by {
                deltas.insert(encode_coord(ax - bx, ay - by));
            }
        }
    }
    deltas
}

fn check_four_vector_prune(selected: &[(i32, i32)], candidate: (i32, i32), deltas: &HashSet<i32>) -> bool {
    if selected.len() < 3 { return true; }
    let (vx, vy) = candidate;
    let sel = selected.len();
    for i in 0..sel - 2 {
        let (ax, ay) = selected[i];
        for j in i + 1..sel - 1 {
            let (bx, by) = selected[j];
            for k in j + 1..sel {
                let (cx, cy) = selected[k];
                for &sa in &[1i32, -1] {
                    let x1 = vx + sa * ax;
                    let y1 = vy + sa * ay;
                    for &sb in &[1i32, -1] {
                        let x2 = x1 + sb * bx;
                        let y2 = y1 + sb * by;
                        let xv = x2 + cx;
                        let yv = y2 + cy;
                        if (xv != 0 || yv != 0) && deltas.contains(&encode_coord(xv, yv)) { return false; }
                        let xv2 = x2 - cx;
                        let yv2 = y2 - cy;
                        if (xv2 != 0 || yv2 != 0) && deltas.contains(&encode_coord(xv2, yv2)) { return false; }
                    }
                }
            }
        }
    }
    true
}

fn build_even_masks(k: usize) -> Vec<usize> {
    let mut masks = Vec::new();
    for mask in 0..(1usize << k) {
        if mask.count_ones() % 2 == 0 { masks.push(mask); }
    }
    masks
}

fn compute_centers(vectors: &[(i32, i32)], masks: &[usize]) -> Vec<(i32, i32)> {
    let mut centers = Vec::new();
    for &mask in masks {
        let mut x = 0i32;
        let mut y = 0i32;
        let mut mm = mask;
        while mm != 0 {
            let lsb = mm & mm.wrapping_neg();
            let idx = lsb.trailing_zeros() as usize;
            x += vectors[idx].0;
            y += vectors[idx].1;
            mm -= lsb;
        }
        centers.push((x, y));
    }
    centers
}

fn quick_harmony_check(centers: &[(i32, i32)], pts: &[(i32, i32)], n: usize) -> bool {
    let mut counts: HashMap<i32, i32> = HashMap::new();
    let mut harmony_count = 0usize;
    for &(cx, cy) in centers {
        for &(vx, vy) in pts {
            let key = encode_coord(cx + vx, cy + vy);
            let cur = counts.entry(key).or_insert(0);
            if *cur == 0 { *cur = 1; }
            else if *cur == 1 { *cur = 2; harmony_count += 1; if harmony_count > n { return false; } }
            else { *cur += 1; }
        }
    }
    harmony_count == n
}

fn strict_check(centers: &[(i32, i32)], pts: &[(i32, i32)], n: usize) -> bool {
    let center_codes: Vec<i32> = centers.iter().map(|&(x, y)| encode_coord(x, y)).collect();
    let center_set: HashSet<i32> = center_codes.iter().cloned().collect();
    let tangent_diffs: Vec<i32> = pts.iter().map(|&(x, y)| encode_coord(2 * x, 2 * y)).collect();
    for &c in &center_codes {
        for &d in &tangent_diffs {
            let other = c + d;
            if center_set.contains(&other) && c < other { return false; }
        }
    }
    let mut pt_to_centers: HashMap<i32, Vec<usize>> = HashMap::new();
    for (idx, &(cx, cy)) in centers.iter().enumerate() {
        for &(vx, vy) in pts {
            let key = encode_coord(cx + vx, cy + vy);
            pt_to_centers.entry(key).or_default().push(idx);
        }
    }
    let harmony_pts: Vec<i32> = pt_to_centers.iter()
        .filter(|(_, v)| v.len() >= 2)
        .map(|(&k, _)| k)
        .collect();
    if harmony_pts.len() != n { return false; }
    let mut parent: Vec<usize> = (0..n).collect();
    let mut sz: Vec<usize> = vec![1; n];
    fn find(parent: &mut Vec<usize>, mut x: usize) -> usize {
        while parent[x] != x {
            parent[x] = parent[parent[x]];
            x = parent[x];
        }
        x
    }
    fn union(parent: &mut Vec<usize>, sz: &mut Vec<usize>, a: usize, b: usize) {
        let ra = find(parent, a);
        let rb = find(parent, b);
        if ra == rb { return; }
        let (ra, rb) = if sz[ra] >= sz[rb] { (ra, rb) } else { (rb, ra) };
        parent[rb] = ra;
        sz[ra] += sz[rb];
    }
    for key in &harmony_pts {
        let lst = &pt_to_centers[key];
        let base = lst[0];
        for j in 1..lst.len() {
            let other = lst[j];
            union(&mut parent, &mut sz, base, other);
        }
    }
    let root = find(&mut parent, 0);
    for i in 1..n {
        if find(&mut parent, i) != root { return false; }
    }
    true
}

fn has_unit_coord(points: &[(i32, i32)]) -> bool {
    for &(x, y) in points {
        if x.abs() == 1 || y.abs() == 1 { return true; }
    }
    false
}

fn dfs(
    pairs: &[((i32, i32), (i32, i32))],
    pts: &[(i32, i32)],
    k: usize,
    masks: &[usize],
    n: usize,
    deltas: &HashSet<i32>,
    selected: &mut Vec<(i32, i32)>,
    start: usize,
) -> bool {
    if selected.len() == k {
        let centers = compute_centers(selected, masks);
        return quick_harmony_check(&centers, pts, n) && strict_check(&centers, pts, n);
    }
    let needed = k - selected.len();
    for pi in start..=pairs.len().saturating_sub(needed) {
        let choices: Vec<(i32, i32)> = if selected.is_empty() {
            vec![pairs[pi].0]
        } else {
            vec![pairs[pi].0, pairs[pi].1]
        };
        for &vec in &choices {
            if !check_four_vector_prune(selected, vec, deltas) { continue; }
            selected.push(vec);
            if dfs(pairs, pts, k, masks, n, deltas, selected, pi + 1) { return true; }
            selected.pop();
        }
    }
    false
}

fn check_valid_oriented_vectors(pairs: &[((i32, i32), (i32, i32))], pts: &[(i32, i32)], k: usize, masks: &[usize], n: usize) -> bool {
    let deltas = build_displacement_set(pts);
    let mut selected: Vec<(i32, i32)> = Vec::new();
    dfs(pairs, pts, k, masks, n, &deltas, &mut selected, 0)
}

fn search_min_radius_sq(k: usize, m_limit: i32, filtered: bool) -> i32 {
    let masks = build_even_masks(k);
    let n = 1usize << (k - 1);
    for m in 1..=m_limit {
        let p = count_antipodal_pairs(m);
        if p < k as i32 { continue; }
        if filtered && p != k as i32 && p != k as i32 + 2 { continue; }
        let pts = circle_points(m);
        if filtered && !has_unit_coord(&pts) { continue; }
        let pairs = compute_opposite_pairs(&pts);
        if pairs.len() != p as usize { continue; }
        if check_valid_oriented_vectors(&pairs, &pts, k, &masks, n) { return m; }
    }
    panic!("No solution found up to m={}", m_limit);
}

fn main() {
    let r1 = search_min_radius_sq(2, 20, false);
    assert_eq!(r1, 1, "Expected 1, got {}", r1);
    let r2 = search_min_radius_sq(3, 50, false);
    assert_eq!(r2, 5, "Expected 5, got {}", r2);
    println!("{}", search_min_radius_sq(10, 20000, true));
}
